// @avail-project/ca-common/dist/esm/_polyfill.js
import * as bufferModule from "buffer";
import { Buffer as DefaultBuffer } from "buffer";
const globalScope = globalThis;
let patched = false;
function collectBufferCtors() {
    const candidates = new Set();
    if (typeof DefaultBuffer === "function") {
        candidates.add(DefaultBuffer);
    }
    const mod = bufferModule;
    if (typeof mod?.Buffer === "function") {
        candidates.add(mod.Buffer);
    }
    const defaultExport = mod?.default;
    if (typeof defaultExport === "function") {
        candidates.add(defaultExport);
    }
    else if (defaultExport &&
        typeof defaultExport === "object" &&
        typeof defaultExport.Buffer === "function") {
        candidates.add(defaultExport.Buffer);
    }
    if (typeof globalScope.Buffer === "function") {
        candidates.add(globalScope.Buffer);
    }
    return Array.from(candidates);
}
function assertOffset(buffer, offset) {
    const numericOffset = Number(offset);
    if (!Number.isFinite(numericOffset)) {
        throw new TypeError("Offset must be a finite number");
    }
    const normalized = numericOffset >>> 0;
    if (normalized !== numericOffset) {
        throw new RangeError("Offset must be a non-negative integer");
    }
    if (normalized + 4 > buffer.length) {
        throw new RangeError("Offset out of bounds");
    }
    return normalized;
}
function fallbackWriteUint32BE(value, offset = 0) {
    const o = assertOffset(this, offset);
    const normalized = Number(value) >>> 0;
    this[o] = (normalized >>> 24) & 0xff;
    this[o + 1] =
        (normalized >>> 16) & 0xff;
    this[o + 2] =
        (normalized >>> 8) & 0xff;
    this[o + 3] = normalized & 0xff;
    return o + 4;
}
function fallbackWriteUint32LE(value, offset = 0) {
    const o = assertOffset(this, offset);
    const normalized = Number(value) >>> 0;
    this[o] = normalized & 0xff;
    this[o + 1] =
        (normalized >>> 8) & 0xff;
    this[o + 2] =
        (normalized >>> 16) & 0xff;
    this[o + 3] =
        (normalized >>> 24) & 0xff;
    return o + 4;
}
function fallbackReadUint32BE(offset = 0) {
    const o = assertOffset(this, offset);
    const store = this;
    return ((store[o] * 0x1000000 +
        ((store[o + 1] << 16) | (store[o + 2] << 8) | store[o + 3])) >>>
        0);
}
function fallbackReadUint32LE(offset = 0) {
    const o = assertOffset(this, offset);
    const store = this;
    return ((store[o] |
        (store[o + 1] << 8) |
        (store[o + 2] << 16) |
        (store[o + 3] * 0x1000000)) >>>
        0);
}
function aliasOrDefine(proto, alias, canonical, fallback) {
    const aliasKey = alias;
    const canonicalKey = canonical;
    const existingAlias = proto[aliasKey];
    if (typeof existingAlias === "function") {
        return;
    }
    const canonicalFn = proto[canonicalKey];
    if (typeof canonicalFn === "function") {
        Object.defineProperty(proto, aliasKey, {
            value: canonicalFn,
            writable: true,
            configurable: true,
        });
        return;
    }
    Object.defineProperty(proto, aliasKey, {
        value: fallback,
        writable: true,
        configurable: true,
    });
}
function patchPrototype(bufferCtor) {
    const proto = bufferCtor.prototype;
    if (!proto) {
        return;
    }
    aliasOrDefine(proto, "writeUint32BE", "writeUInt32BE", fallbackWriteUint32BE);
    aliasOrDefine(proto, "writeUint32LE", "writeUInt32LE", fallbackWriteUint32LE);
    aliasOrDefine(proto, "readUint32BE", "readUInt32BE", fallbackReadUint32BE);
    aliasOrDefine(proto, "readUint32LE", "readUInt32LE", fallbackReadUint32LE);
}
function ensureProcessEnv() {
    if (!globalScope.process) {
        globalScope.process = { env: { NODE_ENV: "production" } };
        return;
    }
    const processValue = globalScope.process;
    if (!processValue.env) {
        processValue.env = { NODE_ENV: "production" };
        return;
    }
    if (typeof processValue.env.NODE_ENV === "undefined") {
        processValue.env.NODE_ENV = "production";
    }
}
export function ensureBufferPolyfill() {
    if (patched) {
        return;
    }
    const candidates = collectBufferCtors();
    candidates.forEach(patchPrototype);
    const preferred = candidates[0];
    if (preferred &&
        (typeof globalScope.Buffer !== "function" ||
            typeof globalScope.Buffer.prototype?.writeUint32BE !==
                "function")) {
        globalScope.Buffer = preferred;
    }
    ensureProcessEnv();
    patched = true;
}
ensureBufferPolyfill();
