// @avail-project/ca-common/dist/cjs/rff/rff.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OmniversalRFF = void 0;
const viem_1 = require("viem");
const data_1 = require("../data");
class OmniversalRFF {
    protobufRFF;
    evmRFF;
    constructor(protobufRFF) {
        this.protobufRFF = protobufRFF;
    }
    asEVMRFF() {
        if (this.evmRFF == null) {
            this.evmRFF = {
                sources: this.protobufRFF.sources.map((s) => ({
                    universe: s.universe,
                    chainID: (0, viem_1.bytesToBigInt)(s.chainID),
                    contractAddress: (0, data_1.ezPadTo32Hex)(s.contractAddress),
                    value: (0, viem_1.bytesToBigInt)(s.value),
                })),
                destinationUniverse: this.protobufRFF.destinationUniverse,
                destinationChainID: (0, viem_1.bytesToBigInt)(this.protobufRFF.destinationChainID),
                recipientAddress: (0, data_1.ezPadTo32Hex)(this.protobufRFF.recipientAddress),
                destinations: this.protobufRFF.destinations.map((d) => ({
                    contractAddress: (0, data_1.ezPadTo32Hex)(d.contractAddress),
                    value: (0, viem_1.bytesToBigInt)(d.value),
                })),
                nonce: (0, viem_1.bytesToBigInt)(this.protobufRFF.nonce),
                expiry: BigInt(this.protobufRFF.expiry.toString()),
                parties: this.protobufRFF.signatureData.map((sd) => ({
                    universe: sd.universe,
                    address_: (0, data_1.ezPadTo32Hex)(sd.address),
                })),
            };
        }
        return this.evmRFF;
    }
    asProtobufRFF() {
        return this.protobufRFF;
    }
}
exports.OmniversalRFF = OmniversalRFF;
