// @avail-project/ca-common/dist/cjs/vaultcontracts/vaultcontracts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Environment = void 0;
exports.getVaultContractMap = getVaultContractMap;
const viem_1 = require("viem");
const data_1 = require("../data");
const definition_1 = require("../proto/definition");
var Environment;
(function (Environment) {
    Environment[Environment["FOLLY"] = 0] = "FOLLY";
    Environment[Environment["CERISE"] = 1] = "CERISE";
    Environment[Environment["CORAL"] = 2] = "CORAL";
    Environment[Environment["JADE"] = 3] = "JADE";
})(Environment || (exports.Environment = Environment = {}));
const dataSets = new Map([
    [
        Environment.FOLLY,
        [
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 421614),
                "0xEFF0C81eC6D7c2a3B924e98B65303DDaa3030a81",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 11155420),
                "0xEFF0C81eC6D7c2a3B924e98B65303DDaa3030a81",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 80002),
                "0xEFF0C81eC6D7c2a3B924e98B65303DDaa3030a81",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 84532),
                "0xEFF0C81eC6D7c2a3B924e98B65303DDaa3030a81",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 11155111),
                "0xEFF0C81eC6D7c2a3B924e98B65303DDaa3030a81",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 10143),
                "0xEFF0C81eC6D7c2a3B924e98B65303DDaa3030a81",
            ],
        ],
    ],
    [
        Environment.CERISE,
        [
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 137),
                "0x111111eA4f8BdfB5AE22c37ebC3eE17b82072F57",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 10),
                "0x111111eA4f8BdfB5AE22c37ebC3eE17b82072F57",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 42161),
                "0x111111eA4f8BdfB5AE22c37ebC3eE17b82072F57",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 8453),
                "0x111111eA4f8BdfB5AE22c37ebC3eE17b82072F57",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 534352),
                "0x111111eA4f8BdfB5AE22c37ebC3eE17b82072F57",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 56),
                "0x111111eA4f8BdfB5AE22c37ebC3eE17b82072F57",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.TRON, 728126428),
                "0x46de8c7e6f1da4dd851b62c20b78971f230fca5b",
            ],
        ],
    ],
    [
        Environment.CORAL,
        [
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 1),
                "0xC0DED5d7F424276c821AF21F68E1e663bC671C3D",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 10),
                "0xC0DED5d7F424276c821AF21F68E1e663bC671C3D",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 137),
                "0xC0DED5d7F424276c821AF21F68E1e663bC671C3D",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 42161),
                "0xC0DED5d7F424276c821AF21F68E1e663bC671C3D",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 534352),
                "0xC0DED5d7F424276c821AF21F68E1e663bC671C3D",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 8453),
                "0xC0DED5d7F424276c821AF21F68E1e663bC671C3D",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 43114),
                "0xC0DED5d7F424276c821AF21F68E1e663bC671C3D",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 999),
                "0xC0DED5d7F424276c821AF21F68E1e663bC671C3D",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 8217),
                "0xC0DED5d7F424276c821AF21F68E1e663bC671C3D",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 50104),
                "0xB61fAdeBccCb15823b64bf47829d32eeb4A08930",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 56),
                "0xC0DED5d7F424276c821AF21F68E1e663bC671C3D",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 143),
                "0xC0DED5d7F424276c821AF21F68E1e663bC671C3D",
            ],
        ],
    ],
    [
        Environment.JADE,
        [
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 1),
                "0x00000000ac0ac9d69424fA5Adc291D75Ec4a0F11",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 10),
                "0x00000000ac0ac9d69424fA5Adc291D75Ec4a0F11",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 137),
                "0x00000000ac0ac9d69424fA5Adc291D75Ec4a0F11",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 42161),
                "0x00000000ac0ac9d69424fA5Adc291D75Ec4a0F11",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 534352),
                "0x00000000ac0ac9d69424fA5Adc291D75Ec4a0F11",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 8453),
                "0x00000000ac0ac9d69424fA5Adc291D75Ec4a0F11",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 43114),
                "0x00000000ac0ac9d69424fA5Adc291D75Ec4a0F11",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 999),
                "0x00000000ac0ac9d69424fA5Adc291D75Ec4a0F11",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 8217),
                "0x00000000ac0ac9d69424fA5Adc291D75Ec4a0F11",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 56),
                "0x00000000ac0ac9d69424fA5Adc291D75Ec4a0F11",
            ],
        ],
    ],
]);
function getVaultContractMap(environment) {
    const src = dataSets.get(environment);
    if (src == null) {
        throw new Error("Environment not found");
    }
    return new data_1.ChainIDKeyedMap(src.map((t) => [
        t[0],
        (0, data_1.zeroExtendBufToGivenSize)((0, viem_1.hexToBytes)(t[1]), 32),
    ]));
}
